examples/secrets-management-with-csi-secrets-driver-aws/main.tf

```diff
@@ -0,0 +1,268 @@
+  create_igw           = true
+  enable_dns_hostnames = true
+  single_nat_gateway   = true
+
+  public_subnet_tags = {
+    "kubernetes.io/cluster/${local.cluster_name}" = "shared"
+    "kubernetes.io/role/elb"                      = "1"
+  }
+
+  private_subnet_tags = {
+    "kubernetes.io/cluster/${local.cluster_name}" = "shared"
+    "kubernetes.io/role/internal-elb"             = "1"
+  }
+}
+#---------------------------------------------------------------
+# Example to consume eks-blueprints module
+#---------------------------------------------------------------
+module "eks-blueprints" {
+  source = "../.."
+
+  tenant            = local.tenant
+  environment       = local.environment
+  zone              = local.zone
+  terraform_version = local.terraform_version
+
+  # EKS Cluster VPC and Subnet mandatory config
+  vpc_id             = module.aws_vpc.vpc_id
+  private_subnet_ids = module.aws_vpc.private_subnets
+
+  # EKS CONTROL PLANE VARIABLES
+  cluster_version = local.cluster_version
+
+  # EKS MANAGED NODE GROUPS
+  managed_node_groups = {
+    mg_4 = {
+      node_group_name = "managed-ondemand"
+      instance_types  = ["m4.large"]
+      min_size        = "2"
+      subnet_ids      = module.aws_vpc.private_subnets
+    }
+  }
+}
+
+module "eks-blueprints-kubernetes-addons" {
+  source         = "../../modules/kubernetes-addons"
+  eks_cluster_id = module.eks-blueprints.eks_cluster_id
+
+  # EKS Managed Add-ons
+  enable_amazon_eks_vpc_cni    = true
+  enable_amazon_eks_coredns    = true
+  enable_amazon_eks_kube_proxy = true
+
+  #K8s Add-ons
+  enable_secrets_store_csi_driver       = true
+  enable_csi_secrets_store_provider_aws = true
+
+  depends_on = [module.eks-blueprints.managed_node_groups]
+}
+
+
+data "aws_partition" "current" {}
+
+data "aws_caller_identity" "current" {}
+
+data "aws_eks_cluster" "eks_cluster" {
+  name = module.eks-blueprints.eks_cluster_id
+}
+
+#---------------------------------------------------------------
+# Parsing the config file to local and also extracting the ARNs of Secret Object
+#---------------------------------------------------------------
+locals {
+  secretconfig   = templatefile("${path.module}/secretconfig.yaml", {})
+  all_secret_arn = [for objects in yamldecode(local.secretconfig) : objects["objectName"]]
+}
+
+#---------------------------------------------------------------
+# Creating IAM Policy to be attached to the IRSA Role
+#---------------------------------------------------------------
+
+resource "aws_iam_policy" "this" {
+  description = "Sample application IAM Policy for IRSA"
+  name        = "${module.eks-blueprints.eks_cluster_id}-${var.application}-irsa"
+  policy      = data.aws_iam_policy_document.secrets_management_policy.json
+}
+
+#---------------------------------------------------------------
+# Creating IAM Role for Service Account
+#---------------------------------------------------------------
+
+module "iam_role_service_account" {
```

Comment:
I believe this can all be removed and the addon should be creating this https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/deec7d5caea47c06dea48fa616ad2c56e52c3cce/modules/kubernetes-addons/helm-addon/main.tf#L61

Is that correct @vara-bonthu ?

Reply:
So, the problem here is that the addon itself has no clue what secrets are being retrieved. The way it works, secrets are retrieved by the AWS provider when a pod needing those secrets is started and removed afterwords. The service-account associated with the respective pods needs to have the needed IRSA config. These policies cannot be added to the addon. In CDK, we do this when the `Team` or `AddOn` secrets are defined. The same thing needs to happen here. That is to say, in order to consume secrets, there are two paths:

1. You need to define your secrets via `Teams` configuration (similar to how we do labels and quotas today and we will have to update the Teams module for this). We need to add the required policies to the IRSA config we define for the Team.
2. For add-ons, this will be a parameter for the add-on in which case we will add the required policies to the addons IRSA config.

CC: @kcaws @shapirov103 I am assuming we want to keep the design for secrets somewhat similar between the two projects.

Reply:
@askulkarni2 / @kcaws - Though i agree on abstracting out IRSA specifics out in a teams configuration, but that work is pretty involved and hence i propose that we proceed with reviewing this PR and merging it and consider the teams configuration with a separate PR.  The implementation details of the module wont change and changes developed in this PR form the basis of next one (+ also provides customers an example on implementation without Teams configuration similar to CDK). I have addressed all the comments otherwise.